airflow/models/dagrun.py

```diff
@@ -311,32 +313,42 @@ def update_state(self, session=None):
 
         leaf_tis = [ti for ti in tis if ti.task_id in {t.task_id for t in dag.leaves}]
 
-        # if all roots finished and at least one failed, the run failed
-        if not unfinished_tasks and any(
-            leaf_ti.state in {State.FAILED, State.UPSTREAM_FAILED} for leaf_ti in leaf_tis
-        ):
-            self.log.info('Marking run %s failed', self)
-            self.set_state(State.FAILED)
-            dag.handle_callback(self, success=False, reason='task_failure',
-                                session=session)
-
-        # if all leafs succeeded and no unfinished tasks, the run succeeded
-        elif not unfinished_tasks and all(
-            leaf_ti.state in {State.SUCCESS, State.SKIPPED} for leaf_ti in leaf_tis
-        ):
-            self.log.info('Marking run %s successful', self)
-            self.set_state(State.SUCCESS)
-            dag.handle_callback(self, success=True, reason='success', session=session)
-
-        # if *all tasks* are deadlocked, the run failed
-        elif (unfinished_tasks and none_depends_on_past and
-              none_task_concurrency and no_dependencies_met):
-            self.log.info('Deadlock; marking run %s failed', self)
-            self.set_state(State.FAILED)
-            dag.handle_callback(self, success=False, reason='all_tasks_deadlocked',
-                                session=session)
-
-        # finally, if the roots aren't done, the dag is still running
+        if conf.getboolean('scheduler', 'REMOVED_TASKS_LEAD_TO_DAGRUN_FAILURE', fallback=False):
+            # REMOVED state counted as unfinished
+            unfinished_tasks = self.get_task_instances(
+                state=State.unfinished_or_removed(),
+                session=session
+            )
```

Comment:
It looks like a trick to me. What do you think to write it directly?
```
            
           removed_tasks_lead_to_dagrun_failure = REMOVED_TASKS_LEAD_TO_DAGRUN_FAILURE
           if removed_tasks_lead_to_dagrun_failure;
              remove_tasks = self.get_task_instances(
                  state=[REMOVED],
                  session=session
              )
             if len(remove_tasks) != 0:
                set failed with descriptive reason
....
```

Reply:
updated